broadcast/broadcast.go

```diff
@@ -0,0 +1,116 @@
+package broadcast
+
+import "context"
+
+// Broadcaster is a communication service with one sender and many recievers with all recievers (subscribers)
+// getting every message sent by the sender. All communication happens via channels.
+type Broadcaster[T any] struct {
+	source         <-chan T
+	listeners      map[chan T]struct{}
+	addListener    chan chan T
+	removeListener chan (<-chan T)
+	bufferSize     int
+	broadcast      func(context.Context, T)
+}
+
+// New creates a Broadcaster that will forward all data from the source channel to subscribers.
+// All subscribers' channels will have the same capacity as source.
+func New[T any](ctx context.Context, source <-chan T) *Broadcaster[T] {
+	return NewBuffered(ctx, source, cap(source), cap(source) == 0)
+}
+
+// NewUnbuffered creates a Broadcaster that will forward all data from the source channel to subscribers.
+// All subscribers' channels will be unbuffered (capacity = 0).
+func NewUnbuffered[T any](ctx context.Context, source <-chan T) *Broadcaster[T] {
+	return NewBuffered(ctx, source, 0, true)
+}
+
+// NewBuffered creates a Broadcaster that will forward all data from the source channel to subscribers.
+// All subscribers' channels will have the capacity of bufferSize. Depending on the value of blockOnSend the Broadcaster
+// will either ensure that all messages are sent to all subscribers (even if that means waiting on unbuffered/full channels)
+// or it can skip any channel whose buffer is either full or 0.
+func NewBuffered[T any](ctx context.Context, source <-chan T, bufferSize int, blockOnSend bool) *Broadcaster[T] {
+	service := &Broadcaster[T]{
+		source:         source,
+		listeners:      make(map[chan T]struct{}),
+		addListener:    make(chan chan T),
+		removeListener: make(chan (<-chan T)),
+		bufferSize:     bufferSize,
+	}
+
+	// Do we want broadcast to block on slow consumers? At least for bufferSize=0 we might have to.
```

Comment:
To reviewers: do you see a better way of handling this? How do we handle both use cases? Do we even try to do it or just fulfill one of these niches?

Reply:
Instead of buffered/unbuffered I'd prefer naming that defines how the messages are delivered, best-effort vs. guaranteed delivery (e.g., Notify vs. Broadcast). In that case I'd also hide the source channel as an implementation detail of the broadcaster, so that we have complete control over buffering etc.

Reply:
Agreed, that naming convention is much nicer.